0000-deb-repo-additional-headers.md

```diff
@@ -0,0 +1,207 @@
+name        string
+created     timestamp
+modified    timestamp
+
+primary key (id)
+```
+
+```
+rhnPackageDebFieldsValues
+==========================
+package_id  number foreign key rhnPackage(id)
+field_id   number foreign key rhnPackageDebFields(id)
+value       string not null
+created     timestamp
+modified    timestamp
+
+primary key (package_id, field_id)
+```
+
+The class `com.redhat.rhn.taskomatic.task.repomd.DebPackageWriter` must be changed to write any additional fields stored in the database into the generated `Packages.gz` file.
+
+## Repo signature checking
+
+In order to import the metadata in a secure way Uyuni should be able to verify the GPG signature of that metadata.
+The GPG verification should be optional but a warning should be shown to the the user in case he's not providing information about the GPG key used to verify the metadata.
+
+### Repo metadata location and signatures
+
+_Note: this section is only informative_
+
+Debian/Ubuntu repos use GPG signing to ensure the integrity of the `Release` file. The signature can be either in a separate file `Release.gpg` or inline in a file called `InRelease`. If both files are present the `InRelease` file is preferred.
+
+Typically the structure of a repo is like this:
+```
+http://mirror.example.com/ubuntu  <- $ARCHIVE_ROOT
+|
++- dists/bionic                   <- Suite or Codename
+   |
+   +- main                        <- $COMPNENT directories
+      |
+      +- binary-amd64
+         |
+         +- Packages.gz           <- Packages indices
+         +- Packages.xz
+         . . .
+      +- binary-i386
+      . . .
+   +- universe
+   . . .
+   +- Release                     <- Plain file
+   +- InRelease                   <- Release file + inline signature
+   +- Release.gpg                 <- Separate signature
+
+```
+
+E.g. for official Ubuntu repos the URL of these files:
+```
+http://mirror.example.com/ubuntu/dists/bionic-updates/InRelease
+http://mirror.example.com/ubuntu/dists/bionic-updates/Release.gpg
+```
+
+There's also an alternative layout for Debian repos called "flat repos". In this case the repo directory containins both `Release*` and `Packages*` files. E.g.:
+```
+https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Master:/Ubuntu1804-Uyuni-Client-Tools/xUbuntu_18.04/
+```
+
+_Note:_
+
+In Uyuni the URL of a deb repository must point to the directory containing the binary packages for a specific architecture. E.g. for Ubuntu 18.04 updates repo is something like:
+```
+http://mirror.example.com/ubuntu/dists/bionic-updates/main/binary-amd64/
+```
+or in case of a flat repository to the directory containing both `Release*` and `Packages*` files:
+```
+https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Master:/Ubuntu1804-Uyuni-Client-Tools/xUbuntu_18.04/
+```
+
+### GPG keys management
+
+The keys used to verify the channel metadata will be stored in the keyring configured in `/etc/rhn/signing.conf` (`GNUPGHOME`)
```

Comment:
I think this is wrong. The signing.conf is only used for signing the outgoing channel metadata.

The keyring used for checking metadata during repo-sync is maintained here: `/var/lib/spacewalk/gpgdir` .
We provide a package `uyuni-build-keys` or `susemanager-build-keys` which contains a default keyring .

There is a default to find GPG keys. When we configure a repository URL we have a checkbox "Has signed Metadata". If this is checked, we expect that the main file has also a `.asc` and a `.key` variant containing the signature and the key.

spacewalk-repo-sync try to check the signature. If the key is not in the keyring it check if the `.key` file is available and ask the user if it is trusted and should be imported.

If the answer is "yes" the key will be imported and the signature could be checked.

It seems that Ubuntu main repos do not have the key at that place.
So we can think about to ship it with our uyuni package. 

Using the fields for Channels maybe tricky as they describe the "outgoing" data, but we need something for the "incomming" data.

Reply:
Good point. I've change the approach to ship Ubuntu (and maybe Debian) keys. For other keys used by custom repos the user can specify the GPG Keyerver (or url) and Key ID at the level of the repository. The key will be imported into `/var/lib/spacewalk/gpgdir/pubring.gpg` which is also used to check Yum repos.